hir-analysis/src/dependency_graph.rs

```diff
@@ -0,0 +1,1449 @@
+                    index: xi,
+                },
+                Self::Conditional {
+                    inst: y_inst,
+                    successor: y_blk,
+                    index: yi,
+                },
+            )
+            | (
+                Self::Conditional {
+                    inst: x_inst,
+                    successor: x_blk,
+                    index: xi,
+                },
+                Self::Indirect {
+                    inst: y_inst,
+                    successor: y_blk,
+                    index: yi,
+                },
+            ) => {
+                let result = x_inst.cmp(y_inst).then(x_blk.cmp(y_blk)).then(xi.cmp(yi));
+                assert_ne!(result, Ordering::Equal, "argument node type conflict");
+                result
+            }
+            (
+                Self::Conditional {
+                    inst: x_inst,
+                    successor: x_blk,
+                    index: xi,
+                },
+                Self::Conditional {
+                    inst: y_inst,
+                    successor: y_blk,
+                    index: yi,
+                },
+            ) => x_inst.cmp(y_inst).then(x_blk.cmp(y_blk)).then(xi.cmp(yi)),
+        }
+    }
+}
+impl PartialOrd for ArgumentNode {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("invalid node identifier")]
+pub struct InvalidNodeIdError;
+
+/// Produce a bit-packed representation of [Node] which is naturally
+/// sortable as if it was the expanded [Node] type.
+///
+/// We currently only need 5 unique values for the node type, so we
+/// use 3 bits, which gives us 7 unique values, thus we have 2 extra
+/// tag values if we ever need them. This leaves us with 61 bits, of
+/// which 32 is reserved for the instruction or value identifier, and
+/// the remaining 29 are available for storing any type-specific data.
+///
+/// We choose a layout that ensures that when compared as an integer,
+/// the sort order of the corresponding [Node] would be identical. This
+/// is a bit tricky, since [ArgumentNode] for example ignores the difference
+/// between Indirect/Conditional argument types when sorted, so in that
+/// case we use the same tag for those types, and differentiate them by
+/// using one of the payload bits as a "conditional" marker. With this
+/// modification, we can place the tag bits first, followed by a
+/// type-specific layout which obeys the ordering rules for that type.
+///
+/// The following is the key for any comments describing bit layout:
+///
+/// * `t`: tag
+/// * `i`: inst
+/// * `v`: value
+/// * `s`: successor
+/// * `x`: index
+/// * `c`: conditional marker
+/// * `0`: unused/zero
+///
+/// ```text,ignore
+/// tttiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiissssssssxxxxxxxx000000000000c
+/// |--tag--|--------payload------------|------------inst--------------|
+/// ```
```

Comment:
I think I meant to remove this, but it got left in for some reason - as you pointed out, the example below this one is correct and has a more thorough explanation.

In any case, the notion of "payload" here is just what I am calling the bits which can contain arbitrary associated data for a given node type, since it varies for most of them. Not all node types have an `inst` (i.e. `hir::Inst`) identifier, some like Result or Stack are tied to a `hir::Value` instead, but all node types have either one or the other (and those values are always encoded using the same number of bits), and I treat that distinctly from "payload" bits for that reason, but that distinction is artificial, since it is not uniform across all payload types.

Reply:
Thanks for the explanation.